File: Algorithms_python/78_Subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        def dfs(i: int):
            if i == len(nums):
                ans.append(t[:])
                return
            dfs(i + 1)
            t.append(nums[i])
            dfs(i + 1)
            t.pop()
        ans = []
        t = []
        dfs(0)
        return ans
'''
dfs深度搜索算法
要同时兼顾数字扩展和位数扩展两个方向，第一个dfs相当于扩展数字，第二个相当于扩展位数
要想通最直接把数字列出来，一看就懂
'''

File: Algorithms_C#/78_Subsets.cs
public class Solution {
    private IList<IList<int>> ans = new List<IList<int>>();
    private List<int> t = new List<int>();

    public IList<IList<int>> Subsets(int[] nums) {
        dfs(0, nums);
        return ans;
    }

    private void dfs(int i, int[] nums){
        if(i == nums.Length){
            ans.Add(new List<int>(t));
            return;
        }
        t.Add(nums[i]);
        dfs(i+1, nums);
        t.RemoveAt(t.Count - 1);
        dfs(i+1, nums);
    }
}